dora/core/common/src/main/java/alluxio/membership/AlluxioEtcdClient.java

```diff
@@ -127,36 +124,41 @@ public static AlluxioEtcdClient getInstance(AlluxioConfiguration conf) {
     return sAlluxioEtcdClient;
   }
 
-  /**
-   * Create jetcd grpc client no forcing.
-   */
-  public void connect() {
-    connect(false);
+  @FunctionalInterface
+  protected interface EtcdUtilCallable<V> {
+    /**
+     * The task where logics to communicate with etcd happens.
+     *
+     * @return Etcd gRPC call result
+     * @throws Exception
+     */
+    V call() throws Exception;
   }
 
-  /**
-   * Create jetcd grpc client and force(or not) connection.
-   * @param force
-   */
-  public void connect(boolean force) {
-    if (mConnected.get() && !force) {
-      return;
-    }
-    mConnected.set(false);
-    // create client using endpoints
-    Client client = Client.builder().endpoints(mEndpoints)
-        .build();
-    if (mConnected.compareAndSet(false, true)) {
-      mClient = client;
+  private <V> V retryInternal(String description, RetryPolicy retryPolicy,
+                              EtcdUtilCallable<V> etcdCallable) throws IOException {
+    Exception ex = null;
+    // TODO(lucy) Currently retry on all sorts of exception and report the last exception,
+    // As jetcd exception often hides underneath CompletableFuture.get(), find
+    // more info later on distinguishing different possible exceptions.
+    while (retryPolicy.attempt()) {
+      try {
+        return etcdCallable.call();
+      } catch (Exception e) {
+        // TODO(lucy) check in future if io.etcd.jetcd.common.exception.EtcdException
+        // needs to be handled for detection of underlying jetcd client connection
+        // getting into possible stale states or needs reconnection, for now, reuse single
+        // jetcd client handle for entire lifecycle of AlluxioEtcdClient.
+        LOG.warn("Failed to {} (attempt {}): {}", description, retryPolicy.getAttemptCount(),
+            e.toString());
+        ex = e;
+      }
+      LOG.debug("AlluxioEtcdClient call failed ({}): ", retryPolicy.getAttemptCount(), ex);
     }
-  }
-
-  /**
-   * Disconnect.
-   * @throws IOException
-   */
-  public void disconnect() throws IOException {
-    close();
+    throw new UnavailableException(
```

Comment:
eventually we would like to throw runtime exception if we fail at the end

Reply:
yeah I'm bit fuzzy on what exception to throw here,  so just referring to the exact layout in AbstractClient.java, let me talk to u offline separately